src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -1,59 +1,225 @@
+    val staleTd = (for (block <- staleBlocks) yield blockQueue.enqueueBlock(block))
+      .lastOption.flatten.map(_.totalDifficulty).getOrElse(0)
+
+    val (executedBlocks, maybeError) = executeBlocks(newBranch)
+    maybeError match {
+      case None =>
+        Right(staleBlocks, executedBlocks)
+
+      case Some(error) =>
+        executedBlocks.foreach { block =>
+          blockchain.removeBlock(block.header.hash)
+          blockQueue.enqueueBlock(block)
+        }
+
+        newBranch.diff(executedBlocks).headOption.foreach { block =>
+          // TODO: implement
+          //blockQueue.removeSubtree(block.header.hash)
+        }
+
+        staleBlocksWithReceiptsAndTDs.foreach { case (block, receipts, td) =>
+          blockchain.save(block)
+          blockchain.save(block.header.hash, receipts)
+          blockchain.save(block.header.hash, td)
+        }
 
-  def executeBlock(
-    block: Block,
-    validators: Validators): Either[BlockExecutionError, Seq[Receipt]] = {
+        Left(error)
+    }
+  }
+
+  // TODO: scaladoc
+  private def executeBlocks(blocks: List[Block]): (List[Block], Option[BlockExecutionError]) = {
+    blocks match {
+      case block :: remainingBlocks =>
+        executeBlock (block) match {
+          case Right (receipts) =>
+            val td = blockchain.getTotalDifficultyByHash(block.header.parentHash).get + block.header.difficulty
+            blockchain.save(block)
+            blockchain.save(block.header.hash, receipts)
+            blockchain.save(block.header.hash, td)
+
+            val (executedBlocks, error) = executeBlocks(remainingBlocks)
+            (block :: executedBlocks, error)
+
+          case Left(error) =>
+          (Nil, Some(error))
+        }
+
+      case Nil =>
+        (Nil, None)
+    }
+  }
+
+  // TODO: scaladoc
+  private def removeBlocksUntil(parent: ByteString, number: BigInt): List[(Block, Seq[Receipt], BigInt)] = {
+    blockchain.getBlockByNumber(number) match {
+      case Some(block) if block.header.hash == parent =>
+        Nil
+
+      case Some(block) =>
+        val receipts = blockchain.getReceiptsByHash(block.header.hash).get
+        val td = blockchain.getTotalDifficultyByHash(block.header.hash).get
+        blockchain.removeBlock(block.header.hash)
+        (block, receipts, td) :: removeBlocksUntil(parent, number - 1)
+
+      case None =>
+        log.error(s"Unexpected missing block number: $number")
+        Nil
+    }
+  }
+
+  // TODO: scaladoc
+  def resolveBranch(headers: Seq[BlockHeader]): BranchResolutionResult = {
+    if (!doHeadersFormChain(headers))
+      InvalidBranch
+    else {
+      val parentByHash = blockchain.getBlockHeaderByHash(headers.head.parentHash)
+
+      parentByHash match {
+        case None =>
+          UnknownBranch
+
+        case Some(parent) =>
+          // find blocks with same numbers in the current chain
+          val oldBranch = getBlocksForHeaders(headers)
```

Comment:
On #305 @ntallar writes:

> Regarding what @LukasGasior1 said, couldn't it be the case that the "old" branch has more blocks than the "new" one and a block is not obtained with getOldBlocks?
Maybe I'm missing something but it seems getOldBlocks only retrieves blocks at the same height as the headers parameter and not at a greater height, could this affect the resolveBranch function?

nb. `getOldBlocks` has been renamed to `getBlocksForHeaders`

Reply:
That's a good point but this whole function is meant only to move the logic previously found in `RegularSync#processBlockHeaders`. 

We would have to compare branches from a common to ancestor down to the chain head. But how do we know we receive all headers that form a parallel branch in the peer?

Reply:
In that case the parallel branch would be the "new" branch? It seems that comparing `from a common to ancestor down to the chain head` is as much as we can do. Also as we are saving blocks from different blockchain branches, if there are better blocks, it will be easier to process them when receiving them later.

However I'm not sure if that's changing the `Ledger#resolveBranch` implementation is needed. Due to how `RegularSync` works, AFAIK the `headers` parameter to `Ledger#resolveBranch` always has greater or equal height than the "old" branch (as it is always equal to `headersQueue`). Maybe it would be better to at least include a comment regarding that.

Reply:
> AFAIK the headers parameter to Ledger#resolveBranch always has greater or equal height than the "old" branch (as it is always equal to headersQueue)

Why is that?

Reply:
I might be missing something, but I think this is caused by the fact that we always start asking for headers (in `RegularSync#askForHeaders`) for the `appStateStorage.getBestBlockNumber() + 1` block, and we accumulate the blocks received in `headersQueue` as we are going "back" in the blockchain (while `resolvingBranches=true`). As this process is not stopped while importing new blocks, the "new" branch will have at least 1 more block than the "old" branch.

I don't like so much about just leaving a comment that this could be easily altered, for example if we enable importing new blocks while syncing.

Reply:
@ntallar I have added a requirement that the oldest header must have number at least equal to current best block number. Otherwise `InvalidBranch` is returned. WDYT?

Reply:
@rtkaczyk That should work :+1:
Having it documented like this would be useful in case any other refactoring is needed :+1: